### PR TITLE
Introduce lucet-validate

### PR DESCRIPTION
**Pull request by pchickey**
Lucet-validate checks a module type (a webassembly module's imports, exports, and their types) against a [witx](https://github.com/WebAssembly/WASI/tree/master/tools/witx) specification.

`lucet-validate` is a library crate, and an executable.

It is optionally usable in `lucetc` as well. If a witx spec is provided to lucetc, it will use lucet-validate to check it as a substitute for wasmparser's wasm validation. lucet-validate uses wasmparser's validator under the hood.

Adds the WASI repo as a submodule, so we have the witx specification in-tree.

Uses witx 0.3.0, which is published on crates.io, and also available via the wasi submodule on the master branch.

### CODE REVIEW

**Comment by pchickey** [comment]
oops, i was trying to see if GH had a way to turn a non-draft back into draft, and apparently not!

**Comment by pchickey** [comment]
Ready for review. I'll probably hold off on merging until https://github.com/WebAssembly/WASI/pull/100 and https://github.com/WebAssembly/WASI/pull/101 get merged, so we can track the WASI repo master.

File: lucet-wasi-sdk/tests/lucetc.rs [hunk 1]
```diff
@@ -117,4 +128,39 @@ mod lucetc_tests {
         */
         let _obj = c.object_file().expect("generate code from c & d");
     }
+
+    #[test]
+    fn hello() {
+        let m = {
+            // This is like module_from_c, except missing the bits that have it leave out the wasi
+            // stdlib and tentry points:
```
Comment (acfoltzer) [inline lucet-wasi-sdk/tests/lucetc.rs]: What bits? Also, `tentry` typo

File: lucet-wasi-sdk/tests/lucetc.rs [hunk 2]
```diff
@@ -117,4 +128,39 @@ mod lucetc_tests {
...
+        let b =
+            Bindings::from_file("../lucet-wasi/bindings.json").expect("load lucet-wasi bindings");
+        let h = HeapSettings::default();
+        let v = Validator::load("../wasi/phases/unstable/witx/wasi_unstable_preview0.witx")
+            .expect("wasi spec validation")
+            .with_wasi_exe(true);
+        let c =
+            Compiler::new(&m, OptLevel::default(), &b, h, false, &Some(v)).expect("compile empty");
+        let mdata = c.module_data().unwrap();
+        assert!(mdata.heap_spec().is_some());
```
Comment (acfoltzer) [inline lucet-wasi-sdk/tests/lucetc.rs]: Could you add some comments that describe what this is testing for at a higher level?

File: lucetc/Cargo.toml [hunk 3]
```diff
@@ -27,6 +27,8 @@ cranelift-faerie = { path = "../cranelift/cranelift-faerie", version = "0.44.0"
 cranelift-wasm = { path = "../cranelift/cranelift-wasm", version = "0.44.0" }
 target-lexicon = "0.8.0"
 lucet-module = { path = "../lucet-module", version = "0.1.1" }
+lucet-validate = { path = "../lucet-validate", version = "0.1.1" }
+witx = { path = "../wasi/tools/witx", version = "0.3.0" }
```
Comment (acfoltzer) [inline lucetc/Cargo.toml]: Unused dependency?

File: lucetc/src/compiler.rs [hunk 4]
```diff
@@ -58,24 +59,30 @@ impl<'a> Compiler<'a> {
...
-                    e.offset,
-                    e.message
-                )
-            })
-            .context(LucetcErrorKind::Validation)?;
+        if let Some(v) = validator {
+            v.validate(wasm_binary)
+                .context(LucetcErrorKind::Validation)?;
+        } else {
+            // As of cranelift-wasm 0.43 which uses wasmparser 0.40, the parser used inside
```
Comment (acfoltzer) [inline lucetc/src/compiler.rs]: wasmparser 0.40 doesn't exist yet?

File: lucet-validate/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+                })
+            } else {
+                Ok(())
+            }
+        } else {
+            Err(Error::ExportNotFound {
+                field: start_name.to_string(),
+            })
+        }
+    }
```
[the remark this answers is not in the document]
Comment (pchickey) [reply lucet-validate/src/lib.rs]: unfortunately, witx isnt yet capable of expressing requirements about export functions, just about the environment a module can import from. We're aware that this is a gap and will figure out some way to solve it.

[the remark this answers is not in the document]
Comment (pchickey) [reply lucet-validate/src/lib.rs]: also, witx is going to support multiple spec files soon, but not quite yet.

File: lucet-wasi-sdk/tests/lucetc.rs (hunk 1 above)
Comment (pchickey) [reply lucet-wasi-sdk/tests/lucetc.rs]: fixed this comment and the other feedback. thanks

File: lucet-validate/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (acfoltzer) [reply lucet-validate/src/lib.rs]: Cool. I'd also love for us to be able to validate that it exports a `malloc` and `free`, probably other things

**Pull request merged**